// trade-republic-api/build/module/lib/tradeRepublicApi.js
import EventEmitter from 'events';
import WebSocket from 'ws';
export class TradeRepublicApi extends EventEmitter {
    constructor(options) {
        super();
        this.subCounter = 1;
        this.locale = '';
        this.sessionToken = '';
        this.sentMessages = new Map();
        this.keepSentMessageHistory = false;
        this.autoReconnect = false;
        this.apiEndpoint = '';
        this.locale = options?.locale || TradeRepublicApi.DEFAULT_LOCALE;
        this.keepSentMessageHistory = options?.keepSentMessageHistory || false;
        this.autoReconnect = options?.autoReconnect || false;
        this.apiEndpoint =
            options?.apiEndpoint || TradeRepublicApi.DEFAULT_APIENDPOINT;
    }
    /**
     * establish a connection to the Trade Republic websocket
     */
    connect() {
        this.ws = new WebSocket(this.apiEndpoint);
        this.ws.on('open', () => {
            this.send(`connect 21 ${JSON.stringify({ locale: this.locale })}`);
        });
        this.ws.on('message', (data) => {
            this.handleMessage(data);
        });
        this.ws.on('close', () => {
            this.emit('close');
            if (this.autoReconnect) {
                this.subCounter = 0;
                this.connect();
                // TODO: re-subscribe active subscriptions?
            }
        });
    }
    /**
     * supply increasing subscription Ids
     * @returns next subscription Id to be used
     */
    getNextSubId() {
        const current = this.subCounter;
        this.subCounter++;
        return current;
    }
    /**
     * subscribe to the ticker topic of isin
     * @param isin isin
     * @param exchangeId exchangeId, only LSX is supported for now
     * @returns subscriptionId
     */
    subTicker(isin, exchangeId = 'LSX') {
        return this.sub('ticker', { id: `${isin}.${exchangeId}` });
    }
    /**
     * subscribe to a topic by type and specify payload
     * @param type message type
     * @param payload message payload
     * @returns subscriptionId
     */
    sub(type, payload) {
        const subId = this.getNextSubId();
        const msg = `sub ${subId} ${JSON.stringify({
            type,
            token: this.sessionToken,
            ...payload,
        })}`;
        this.sentMessages.set(subId, { type, payload });
        this.send(msg);
        return subId;
    }
    /**
     * subscribe to a topic and wait for a single response
     * @param type message type
     * @param payload message payload
     * @returns response data
     */
    async oneShot(type, payload) {
        const subId = this.sub(type, payload);
        return new Promise((resolve) => {
            const func = (data) => {
                resolve(data);
                this.removeListener(`sub#${subId}`, func);
                this.send(`unsub ${subId}`);
            };
            this.on(`sub#${subId}`, func);
        });
    }
    /**
     * emit an event only when they are listened for
     * @param eventName name of the event
     * @param data payload of the event
     */
    emitIfSubbed(eventName, data) {
        if (this.eventNames().indexOf(eventName) > -1) {
            this.emit(eventName, data);
        }
    }
    /**
     * get all messages that have been sent to the websocket
     * empty if `keepSentMessageHistory` is not enabled
     * @returns map linking subscription Ids to request payloads
     */
    getSentMessages() {
        return this.sentMessages;
    }
    /**
     * process an incoming message
     * @param data raw received websocket message
     */
    handleMessage(data) {
        this.emitIfSubbed('data', data);
        if (data == TradeRepublicApi.CONNECTED_CONFIRMED_MSG) {
            this.emit('connected');
        }
        else {
            try {
                const parsed = this.processNewMessage(data.toString());
                const { subId } = parsed;
                const request = this.sentMessages.get(subId);
                if (!this.keepSentMessageHistory) {
                    this.sentMessages.delete(subId);
                }
                this.emitIfSubbed(`sub#${subId}`, { request, ...parsed });
                // TODO: better ticker filter
                if (data.toString().indexOf('bid":') > -1) {
                    this.emitIfSubbed('ticker', { request, ...parsed });
                }
                else if (data.toString().indexOf('expectedClosingTime') > -1) {
                    this.emitIfSubbed('historical', { request, ...parsed });
                }
            }
            catch (error) {
                this.emit(`error`, { error, data });
            }
        }
    }
    /**
     * sends a message to the Websocket
     * @param msg message string
     */
    send(msg) {
        this.ws.send(msg);
        this.emitIfSubbed('sent', msg);
    }
    /**
     * extracts structured data from an incoming message
     * @param data raw received data
     * @returns Message obj containing subId, type and payload
     */
    processNewMessage(data) {
        const parts = data.split(' ');
        const subId = parseInt(parts[0]);
        const type = parts[1];
        const json = parts[2];
        const payload = JSON.parse(json);
        return {
            subId,
            type,
            payload,
        };
    }
}
TradeRepublicApi.DEFAULT_LOCALE = 'de';
TradeRepublicApi.DEFAULT_APIENDPOINT = 'wss://api.traderepublic.com';
TradeRepublicApi.CONNECTED_CONFIRMED_MSG = 'connected';
